packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,68 @@
+import debugFactory from 'debug';
+import { playQueuedActions } from './shopping-cart-reducer';
+import type { ShoppingCartState, ShoppingCartReducerDispatch, CacheStatus } from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+function fetchInitialCart(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' && cacheStatus !== lastCacheStatus ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+		dispatch( { type: 'GET_CART_FROM_SERVER' } );
+	}
+}
+
+function prepareInvalidCartForSync(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { queuedActions, cacheStatus } = state;
+	if (
+		queuedActions.length === 0 &&
+		cacheStatus === 'invalid' &&
+		cacheStatus !== lastCacheStatus
+	) {
+		debug( 'triggering sync of cart to server' );
+		dispatch( { type: 'REQUEST_UPDATED_RESPONSE_CART' } );
+		dispatch( { type: 'SYNC_CART_TO_SERVER' } );
```

Comment:
That's how this used to work before https://github.com/Automattic/wp-calypso/pull/54300 but it caused possible race conditions. By always using the dispatcher, we guarantee that actions are always queued in order. These two lines say, "if the cart needs to be synced, queue two actions: the first one to lock the cart to prevent further changes, and the second to perform the sync". See the examples in the above PR for a more detailed description of how we use the JS event loop as an action queue.

Reply:
This should be resolved now as I modified the dispatcher to be synchronous (except for the state based actions).

Reply:
Yes, doing the dispatches synchronously while deferring the lock+sync to a next event loop tick should indeed solve the #54300 race condition and allow convenient batching :+1: